src/bin_pattern.jl

```diff
@@ -0,0 +1,169 @@
+"""
+    bin_sparse_matrix!(M::AbstractArray, M_id::AbstractSparseMatrixCSC, max_num_bins::Integer)
+
+Compute a Binning pattern in `M_id` from the sparsity pattern in `M_id`.
+
+`M_id` is SparseMatrixCSC{Int64, Int64} with the same shape as `M`,
+when handed over it contains only 0 and 1. After usage it contains an Integer
+corresponding to a bin or 0. According to the Binning Pattern,
+similar valued entries can be considered the same.
+
+`M` is the matrix whose entries are used for binning. `max_num_bins` is the maximum
+number of bins and must be non-negative, if `max_num_bins=0`, no binning is
+performed and only the sparsity pattern in `M_id` is used then.
+
+See also: [`sps_compute`](@ref), [`p_norm_sparsity_matrix`](@ref).
+
+# Examples
+```jldoctest
+julia> bin_sparse_matrix!([4 1 4.01; 0.1 17.1 17; 0.2 4 29], sparse([1 0 1; 0 1 1; 0 0 1]), 200)
+3×3 SparseMatrixCSC{Int64, Int64} with 5 stored entries:
+ 1  ⋅  1
+ ⋅  2  2
+ ⋅  ⋅  3
+```
+"""
+function bin_sparse_matrix!(M::AbstractArray{T}, M_id::AbstractSparseMatrixCSC, max_num_bins::Integer) where {T}
+    @assert max_num_bins >= 0 "max_num_bins must be non-negative"
+    @assert size(M, 1) == size(M_id, 1) && size(M, 2) == size(M_id, 2) "M must be the same size as M_id"
+
+    if T <: Integer
+        M = 1.0*M
+    end
+    P = eltype(M)
+
+    m, n = size(M_id)
+    M_i, M_j = findnz(M_id)
+    M_pat_nnz = count(!iszero, M_id)
+
+    separated_at = zero(P)
+
+    if max_num_bins == 0
+        data = Vector(1:M_pat_nnz)
+        M_id[:, :] = sparse(M_i, M_j, data, m, n)
+    else
+        nz_M = zeros(M_pat_nnz)
```

Comment:
Might be better to specify the type of `zeros`, e.g. `zeros(T, M_pat_nnz)` (not sure this is correct).